### PR TITLE
After all, why WOULDN'T I add more jank?!

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
The current implementation of TXN3 works, but it is janky. Let's replace it with different jank.

**Implementation Description (bullets)**:
Old implementation would put the committed value at an atlas and cassandra timestamp Long.MAX_VALUE - 10. This works because we are reading at atlas timestamp MAX_VALUE, but it leaves the old entry there forever, and is also really messy for the restore from backup workflow.

Now we have a new KVS method that lets us overwrite PuE and CaS writes, by writing at atlas timestamp 0 and cassandra timestamp as above. This is not great because implementing it for other KVSs is not obvious, but we can leave that for later?

**Testing (What was existing testing like?  What have you done to improve it?)**:
I added an integration test

**Concerns (what feedback would you like?)**:
Is this too jank? Do we worry about what happens if we try to CaS the value that was put there? This does not happen in our protocol, but seems janky to have a KVS method that behaves in undefined ways when the javadoc is not followed precisely.

**Where should we start reviewing?**:

**Priority (whenever / two weeks / yesterday)**:

### CODE REVIEW

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/KvsConsensusForgettingStoreTest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    private final InMemoryKeyValueService kvs = new InMemoryKeyValueService(true);
+    ConsensusForgettingStore store = new KvsConsensusForgettingStore(kvs, TABLE);
+
+    @Test
+    public void putOverwritesPuE() throws ExecutionException, InterruptedException {
+        store.putUnlessExists(CELL, SAD);
+        store.put(CELL, HAPPY);
+        assertThat(store.get(CELL).get().get()).containsExactly(HAPPY);
+        assertThat(kvs.getAllTimestamps(TABLE, ImmutableSet.of(CELL), Long.MAX_VALUE)
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/KvsConsensusForgettingStoreTest.java:43]: there's a hasSize() annotation I think?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java [hunk 2]
```diff
@@ -393,42 +393,70 @@ public void multiPut(Map<TableReference, ? extends Map<Cell, byte[]>> valuesByTa
...
             }
-            knownSuccessfullyCommittedKeys.add(entry.getKey());
         }
     }
 
+    private boolean shouldThrow(OverwriteBehaviour overwriteBehaviour, byte[] contents, byte[] oldContents) {
+        return oldContents != null
+                && (overwriteBehaviour == OverwriteBehaviour.DO_NOT_OVERWRITE
+                        || (overwriteBehaviour == OverwriteBehaviour.OVERWRITE_SAME_VALUE
+                                && !Arrays.equals(oldContents, contents)));
```
Comment (jeremyk-91) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java:451]: For readability, could we separate this out into terms? I almost prefer the less concise, if more direct approach:

```
if empty return false
switch (OB) {
  case DO_NOT_OVERWRITE: return true;
  case OVERWRITE_SAME_VALUE: return !Arrays.equals(...);
  case OVERWRITE: return false;
  default: throw new IllegalStateException();
}
```

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellValuePutter.java [hunk 3]
```diff
@@ -61,18 +70,22 @@ public CellValuePutter(
...
-        put(kvsMethodName, tableRef, values, false);
+        putInternal(kvsMethodName, tableRef, values, Optional.empty());
     }
 
-    private void put(
+    void set(final String kvsMethodName, final TableReference tableRef, final Iterable<Map.Entry<Cell, Value>> values) {
+        putInternal(kvsMethodName, tableRef, values, Optional.of(SET_TIMESTAMP));
+    }
+
+    private void putInternal(
```
Comment (jeremyk-91) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellValuePutter.java:84]: It is internal, but might be useful to have docs to explain how `overrideTimestamp` works

**Review by jeremyk-91 — APPROVED** [review]
Agree that this, while janky, is less janky than the original approach (and probably more in line with the standard conventions for atomic tables: there is one timestamp value for each Cell that is updated).

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/KvsConsensusForgettingStoreTest.java (hunk 1 above)
Comment (gmaretic) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/KvsConsensusForgettingStoreTest.java:43]: not for Multimap :(

**Pull request merged**